src/viam/examples/camera/example_camera.cpp

```diff
@@ -54,9 +54,9 @@ int main() {
         std::string camera_name("camera1");
 
         cout << "Getting camera: " << camera_name << endl;
-        std::shared_ptr<vs::CameraClient> camera;
+        std::shared_ptr<vs::Camera> camera;
         try {
-            camera = robot->resource_by_name<vs::CameraClient>(camera_name);
+            camera = robot->resource_by_name<vs::Camera>(camera_name);
```

Comment:
This is an example of the change necessitated by option (1) in the note to reviewers. Unless we add an additional templated `api_map` struct to the `CameraClient` class, `resource_by_name<CameraClient>(...)` will fail to work because `resource_by_name` needs to get the `CameraClient` API.

Reply:
FWIW I don't think stating the templated type as `Camera` here is too crazy. In the Go SDK, `RobotClient.ResourceByName` returns a `resource.Resource`, even though the value underlying that interface that's actually returned is a resource _client_.

Reply:
I think there is an interesting question here. Because the effect of this change is wider than just whether `resource_by_name<FooClient>` works vs `resource_by_name<Foo>`. The former is going to return you a downcasted pointer with static type `FooClient`, and the latter will have static type `Foo`. It means that you can no longer really access the `*Client` types directly. Are they even still part of the public API of the SDK? Or are they an implementation detail now and all interaction is mediated via the `Foo` type, even though the dynamic type of the pointed to object is actually a `FooClient` object?

If the answer is that `FooClient` types are now an implementation detail, then I sort of like that the result here is that it forces us to talk about things via the resource type, not the concrete type. If `FooClient` types are *not* an implementation detail, what might they be needed for by an end-user of the SDK that they can't get by using `Foo` instead?

If we do think it is important that it be possible to get a resource with the `FooClient` static type instead of only `Foo`, I think there is a way to make both work, without too much boilerplate, by adding a `resource_type` typedef to `FooClient` and updating `resource_by_name` with some SINFAE magic to make it work for both `FooClient` and `Foo` depending on whether a `resource_type` is present. I'd need to check to make sure that can actually be done but I'm fairly confident it could.

Either way, I'm happy that this review is forcing us to decide the issue of whether the `FooClient` types are actually user-facing or an implementation detail.

Reply:
> If the answer is that `FooClient` types are now an implementation detail

I like `FooClient` becoming hidden from SDK users; that's effectively what happens in the Go SDK. 

For more detail: There is a `resource.Resource` interface embedded in another interface like `motor.Motor`. `motor.Motor` is implemented by actual motors _and_ by `motor.MotorClient`. When an SDK user calls `motor.FromRobot(r, "mymotor")`, an object implementing `motor.Motor` is returned, despite the actual type of that object being `motor.MotorClient`. In that way, `motor.MotorClient` is hidden from SDK users.

Reply:
Has me wondering about the `FooServer` types. Maybe they too don't actually form part of the public SDK API.

Reply:
Yeah, interesting. The `MLModelServiceServer` type does not appear in the triton server module at all.

Reply:
Yeah the more I think about it the more it seems that both the `Server` and `Client` types are implementation details that users don't really need to know about/interact with. So I think I'm pretty happy with changing the `resource_by_name` call pattern from being `<FooClient>` to `<Foo>`.